vaadin-upload-flow-parent/vaadin-upload-flow/src/main/java/com/vaadin/flow/component/upload/Upload.java

```diff
@@ -493,6 +505,517 @@ protected void fireUpdateProgress(long totalBytes, long contentLength) {
+            ComponentEventListener<UploadRetryEvent> listener) {
+        return addListener(UploadRetryEvent.class,
+                (ComponentEventListener) listener);
+    }
+
+    @DomEvent("upload-start")
+    public static class UploadStartEvent extends ComponentEvent<Upload> {
+        private final JsonObject detail;
+        private final JsonObject detailXhr;
+        private final JsonObject detailFile;
+
+        public UploadStartEvent(Upload source, boolean fromClient,
+                @EventData("event.detail") JsonObject detail,
+                @EventData("event.detail.xhr") JsonObject detailXhr,
+                @EventData("event.detail.file") JsonObject detailFile) {
+            super(source, fromClient);
+            this.detail = detail;
+            this.detailXhr = detailXhr;
+            this.detailFile = detailFile;
+        }
+
+        public JsonObject getDetail() {
+            return detail;
+        }
+
+        public JsonObject getDetailXhr() {
+            return detailXhr;
+        }
+
+        public JsonObject getDetailFile() {
+            return detailFile;
+        }
+    }
+
+    /**
+     * Adds a listener for {@code upload-start} events fired by the
+     * webcomponent.
+     *
+     * @param listener
+     *            the listener
+     * @return a {@link Registration} for removing the event listener
+     */
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    protected Registration addUploadStartListener(
+            ComponentEventListener<UploadStartEvent> listener) {
+        return addListener(UploadStartEvent.class,
+                (ComponentEventListener) listener);
+    }
+
+    @DomEvent("upload-success")
+    public static class UploadSuccessEvent extends ComponentEvent<Upload> {
+        private final JsonObject detail;
+        private final JsonObject detailXhr;
+        private final JsonObject detailFile;
+
+        public UploadSuccessEvent(Upload source, boolean fromClient,
+                @EventData("event.detail") JsonObject detail,
+                @EventData("event.detail.xhr") JsonObject detailXhr,
+                @EventData("event.detail.file") JsonObject detailFile) {
+            super(source, fromClient);
+            this.detail = detail;
+            this.detailXhr = detailXhr;
+            this.detailFile = detailFile;
+        }
+
+        public JsonObject getDetail() {
+            return detail;
+        }
+
+        public JsonObject getDetailXhr() {
+            return detailXhr;
+        }
+
+        public JsonObject getDetailFile() {
+            return detailFile;
+        }
+    }
+
+    /**
+     * Adds a listener for {@code upload-success} events fired by the
+     * webcomponent.
+     *
+     * @param listener
+     *            the listener
+     * @return a {@link Registration} for removing the event listener
+     */
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    protected Registration addUploadSuccessListener(
```

Comment:
Could be `private` since it's only used internally

Reply:
Removed as suggested by https://github.com/vaadin/flow-components/pull/4364#discussion_r1050476577